### PR TITLE
[Attn] Add Parallax (parameterized local linear attention) op, layer and model

### PR DESCRIPTION
**Pull request by Yifei-Zuo**
## Summary
Parallax is an attention variant that augments softmax attention with a learned **first-order correction** computed from a second query-side stream `r`. In addition to the usual scores `S1 = scale·(QK^\top)`, it forms an unscaled `S2 = RK^\top` and reweights the softmax probabilities by it, yielding `o = O1/d1·(1 + d2/d1) − O2/d1` (where `O1, O2` are the value aggregations under `P1 = softmax(S1)` and `P2 = P1·S2`). This adds a content-dependent, parameterized local term on top of standard attention.

See the [paper](https://arxiv.org/abs/2605.29157) and [github](https://github.com/Yifei-Zuo/Parallax) for details.

This PR introduces the Parallax training and decode kernels and the Parallax model. Concretely:
1. **Reference implementation** in PyTorch (`naive_parallax`), fp32, for pedagogical and testing purposes.
2. **Training kernel** in Triton, with GQA, sliding window, and variable-length support.¹
3. **Decode kernels** in Triton.¹
4. **A `Parallax` layer** with q/r/k/v/o projections, optional QK-norm, GQA, sliding window, and KV-cache integration; routes prefill/training through the training kernel and cached decode through the decode kernels.
5. **A `ParallaxModel` / `ParallaxForCausalLM`** auto-registered with the Transformers library.

*¹ The Triton kernels are currently developed for H100 GPUs.*

## Test Plan
- `tests/ops/test_parallax.py`: forward / backward (dq, dr, dk, dv) parity vs. the PyTorch reference, across GQA, sliding window, and variable-length inputs, plus decode self-consistency (`decode` and `decode_onestep` vs. the training kernel).
- `tests/models/test_modeling_parallax.py`: forward / backward and generation smoke tests for `ParallaxForCausalLM`.
- Variable-length and generation are wired and asserted, so `ParallaxConfig` is not added to `MODELING_UNSUPPORTED_VARLEN` / `GENERATION_UNSUPPORTED`.
- Tolerances follow the existing `assert_close` convention.

## Breaking changes
None. This PR only adds new files and registrations; no existing APIs change.

### CODE REVIEW

File: fla/models/parallax/modeling_parallax.py [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+        if labels is not None:
+            if getattr(self, 'criterion', None) is None:
+                if self.config.fuse_linear_cross_entropy:
+                    criterion = FusedLinearCrossEntropyLoss(use_l2warp=self.config.use_l2warp)
+                elif self.config.fuse_cross_entropy:
+                    criterion = FusedCrossEntropyLoss(inplace_backward=True)
+                else:
+                    criterion = nn.CrossEntropyLoss()
+            else:
+                criterion = self.criterion
```
[the remark this answers is not in the document]
Comment (Yifei-Zuo) [reply fla/models/parallax/modeling_parallax.py:409]: This is a established pattern in every model that defines an inline criterion, none of which write back to `self.criterion`. A few current references:

- GLA — [`modeling_gla.py` L353 + L430-438](https://github.com/fla-org/flash-linear-attention/blob/7378dfed2ac08242b3770be43fd4539df636afc3/fla/models/gla/modeling_gla.py#L430-L438)
- Wall Transformer (merged 2026-06-03) — [`modeling_wall_transformer.py` L369 + L431-439](https://github.com/fla-org/flash-linear-attention/blob/7378dfed2ac08242b3770be43fd4539df636afc3/fla/models/wall_transformer/modeling_wall_transformer.py#L431-L439)
- MoBA — [`modeling_moba.py` L321 + L397-405](https://github.com/fla-org/flash-linear-attention/blob/7378dfed2ac08242b3770be43fd4539df636afc3/fla/models/moba/modeling_moba.py#L397-L405)

All use `self.criterion = None` in `__init__` and a local `criterion` in `forward`.

Happy to revisit repo-wide if maintainers prefer real caching.

**Comment by yzhangcs** [comment]
@Yifei-Zuo Great job! will take a look at it soon.

**Comment by yzhangcs** [comment]
@codex review

**Review by yzhangcs — APPROVED** [review]
